### ISSUE TITLE
Issue with Openmpi compilation on case-insensitive systems with clang

### ISSUE THREAD

**Issue by srpgilles** [open-mpi/ompi#9122]
### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

Tried with 4.0.5, 4.0.6, 4.1.0 and 4.1.1.

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

From a tarball, with command

```shell
 ./configure --enable-cxx-exceptions --with-hwloc=internal --with-libevent=internal  --prefix=/Volumes/Data/ci/opt/clang_release/Openmpi CC=clang CXX=clang++ FC=/usr/local/Cellar/gcc/11.1.0_1/bin/gfortran CFLAGS='-m64  -O3 -DNDEBUG ' CXXFLAGS='-m64  -stdlib=libc++ -O3 -DNDEBUG ' FCFLAGS='-m64   -O3 -DNDEBUG ' LDFLAGS='-m64  -O3 -DNDEBUG ' --enable-mpi-cxx
```

### Please describe the system on which you are running

Tried both on macOS Catalina and macOS BigSur; computers were few years old MacPro.

## Details of the problem

When compiling Openmpi, I get messages like:

```shell
In file included from comm.cc:26:
In file included from ../../../ompi/mpi/cxx/mpicxx.h:42:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/map:481:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__tree:14:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/iterator:419:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__functional_base:14:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/type_traits:417:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/cstddef:37:
../../../version:1:3: error: invalid preprocessing directive
# Copyright (c) 2008      Sun Microsystems, Inc.  All rights reserved.
  ^
../../../version:2:3: error: invalid preprocessing directive
# Copyright (c) 2008-2019 Cisco Systems, Inc.  All rights reserved.
  ^
../../../version:3:3: error: invalid preprocessing directive
# Copyright (c) 2011      NVIDIA Corporation.  All rights reserved.
  ^
../../../version:4:3: error: invalid preprocessing directive
# Copyright (c) 2013      Mellanox Technologies, Inc.
  ^
../../../version:5:27: error: invalid preprocessing directive
#                         All rights reserved.
                          ^
../../../version:6:3: error: invalid preprocessing directive
In file included from file.cc# Copyright (c) 2017      Los Alamos National Security, LLC.  All rights:
18:
  ^In file included from 
../../../ompi/mpi/cxx/mpicxx.h:42:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/map../../../version::4817:
:In file included from 27/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__tree:: 14:
```

I was a bit astonished, as compiling on my MacBookPro with BigSur went fine... After investigating a bit I think this is the same issue [reported here in another project](https://github.com/spack/spack/issues/11515) (I don't know if they mentioned it to you or not).

In particular, I think [this comment](https://github.com/spack/spack/issues/11515#issuecomment-495626063) nails it: the error message is really the first lines of the _VERSION_ file, and looking into _cstddef_ there is an `#include <version>` which is in all likelihood poorly interpreted... This would explain why I wasn't stuck on my personal computer: I chose a case sensitive filesystem, contrary to both MacPro on which I tried to compile Openmpi later.

The comment cited above suggests it is because _stage_ directory is put before the system includes, so the fix may not be difficult to provide (but I must admit I'm not proficient with autotools...).

**Comment by hjelmn** [comment]
Does this go away if you disable the C++ bindings? The bindings were removed from the standard 9 years ago now. They will not be in Open MPI 5.0.0.

**Comment by srpgilles** [comment]
Yes you're right! (and I even wasn't using those bindings...)

Many thanks!

**Comment by jsquyres** [comment]
This is actually a real issue.  We don't configure/compile the C++ bindings by default in v4.0.x and v4.1.x, but they're still there, and so they should still work (i.e., it's a regression if they don't work).

I have a workaround that I'll PR shortly.

**Comment by jsquyres** [comment]
This has been merged in both the v4.0.x and v4.1.x branches, and will be included in the next releases of both.

Thanks!
